### ISSUE TITLE
[roscpp] Parameter caching doesn't work correctly for namespaces

### ISSUE THREAD

**Issue by RainCT** [ros/ros_comm#351]
Currently using getParamCached on a namespace results in:
 a) The parameter not being updated correctly when one of its childs changes.
 b) A memory leak where the value of the child is blindly stored into the g_params map (and never removed).

I'll submit a patch.

**Comment by dirk-thomas** [comment]
Can you please implement a unit test for this issue which fails before and passes after your patch?

**Comment by RainCT** [comment]
Sure, I've added one to my pull request.

**Comment by RainCT** [comment]
Do you have any more feedback on this?

**Comment by dirk-thomas** [comment]
The changes PR #352 does are pretty significant. It combines bug fixes (the missing invalidation of the cached parent parameters) with new features (adding caching of non-existing parameters) which makes it more difficult to judge the impact and potential issues.

I have separated the bug fix into a separate PR which should address **a)** of your original comment: #384 

Can you please point out where exactly the memory leak is which you mentioned in **b)**?

**Comment by RainCT** [comment]
"g_params[clean_key] = v;" in update() is adding a new key to the cache dictionary, which isn't tracked anywhere  (note that if you subscribe to X and then store a tree in X, update will be called with X/foo whenever a child foo of X changes).

**Comment by dirk-thomas** [comment]
`g_params` stores the cached parameters until they are `del()`eted. I don't understand what you mean with "isn't tracked anywhere" and why that would be a memory leak.

Regarding the update notification: based on your short description I would say it is intended that the subscriber of X continues to be notified. May be I don't get your use case correctly. Could you create a unit test for this (which fails currently and passes with the patch proposed in this PR)?

**Comment by RainCT** [comment]
$ rosparam set /foo 'hi'
c++: Code subscribes to /foo

$ rosparam set /foo '{sub: 1}'
c++: updates g_params['/foo'], ok
$ rosparam set /foo/sub 2
c++: since you are subscribed to /foo, update('/foo/sub', ...) is called: it creates g_params['/foo/sub'] (which will never be accessed)
$ rosparam delete /foo
c++: g_params['/foo'] is erased, but g_params['/foo/sub'] is leaked

IIRC, with my patch the behaviour will be:
  a) if update('/foo', ...) is called, it is updated
  b) if update('/foo/sub', ...) is called, updateNamespaceImpl will delete g_params['/foo'], so next time it will be queried again (alternatively g_params['/foo'] could be updated, but that would require messing with the cached XmlRpc instance). NO g_params['/foo/sub']  entry will be created.

[A special case in a) is deletion, since it's not possible to distinguish between "delete /foo" and "set foo {}"]

**Comment by dirk-thomas** [comment]
I don't think that this is fixable cleanly with the current API. The fact that you can get a whole subtree is not consistent in the rosparam API. E.g. after deleting `/foo` in your example it is completely valid to get `/foo/sub` afterwards (and you actually should be able to). Therefore I would _not_ consider this a memory leak.

The problem is that as soon as you treat `/foo` as a subtree the whole current caching logic does not work, e.g. setting `/foo` to a value which contains a subtree does not work either since it does not affect the sub elements in any way.

Therefore I would propose to address the case where the parent cache entries are invalidated with #384. And not address the other cases for now. That would require a significant update to the rosparam caching logic. While that should be addressed in a future redesign I don't think it is feasible to fix in indigo - also since it would change the current behavior of the API.

**Comment by RainCT** [comment]
> to get /foo/sub afterwards

If you request /foo/sub (while it still exists), it will NOT be taken from the cache, since it isn't in g_subscribed_params. However, it'll remain in g_params forever, even when it gets deleted.

(Well, if you request it, then it'll be queried, added to g_subscribed_params and the entry in g_params will be overriden, from that point all is fine. But if you never request a child that's been added to g_params, it's leaked.)

> E.g. after deleting /foo in your example it is completely valid to get /foo/sub afterwards (and you actually should be able to).

Afaics that's not how it works:

siegfriedgevatter@thailand:svn:
~$ rosparam set /foo '{}'
siegfriedgevatter@thailand:svn:
~$ rosparam set /foo/sub 5
siegfriedgevatter@thailand:svn:
~$ rosparam delete /foo
siegfriedgevatter@thailand:svn:
~$ rosparam get /foo/sub
ERROR: Parameter [/foo/sub] is not set

**Comment by dirk-thomas** [comment]
For the second part I would have expected the second parameters to be still there - at least when the first one was set to a primitive value. But that is indeed not the case.

I think I got the first point. That is the reason why you remove the nested iterator loops in the `del()` function. I have added a3d32ec92c461ca168f5f44eca4efbbda909798f to the PR (just for completeness). Would you think that this covers the case?

Could you please also describe (again, sorry) why you think the additional changes in #352 are necessary / should be merged over what is currently in #384?

**Comment by RainCT** [comment]
Hi,

Sorry if the description wasn't clear enough, I understand that it takes a while to get ones head around param.cpp.

Just to clarify: in roscore, params are stored as a tree, where only leaf nodes have a value. When you request a param, it'll look it up in the tree: if it's a leaf node, the value is returned; if it's not, then the subtree starting at that node is returned.

The caching code isn't aware of this fact, which leads to two problems:
 a) the memory leak (which is rather simple to fix)
 b) that if a non-leaf node is requested and a child was modified, you won't see the change

As for this second issue, my branch fixes it by deleting the cached entry when a child of it changes. On the subsequent getParamCached request for that node, it'll be queried again, and thus will return the correct value.

This fix is just a compromise to stop it from being entirely broken (for non-leaf nodes). A better fix would be to replicate the tree structure in the cache, but that'd be a more significant code change.

## 

Regarding the introduction of g_nonexistent_params, that's to remember whether a param exists (and not request it again each time getParamCached is called for it). I don't remember the details of why I did it like that.

I'll try to look into it tomorrow, split the stuff up into smaller merge commits, and add some more test case / comment.

## 

By the way, thanks for looking into this :).

**Comment by dirk-thomas** [comment]
That sounds like that the much smaller PR #384 covers both cases a) as well as b). The introduction of `g_nonexistent_params` serves the purpose to improve performance for known non-existing parameters which would be nice-to-have but is not necessary and could be left out to not increase the code complexity even further. I think if we add a unit test covering a) to #384 that one should be good to go?

And I completely agree with you that the general design problem is more challenging to fix. Therefore I would leave that to a redesign of ROS parameters rather then trying to address it in _Indigo_.

**Comment by dirk-thomas** [comment]
Since there was no further response I assume #384 is an adequate replacement. Closing this one.

**Comment by RainCT** [comment]
Hi Dirk. Sorry for the delay.

It looks like you didn't commit #384 yet.

Also, it's still missing the fix for the leak. Changing the following in the update() function:

``` c++
g_params[clean_key] = v;
```

to:

``` c++
if (g_subscribed_params.find(clean_key) != g_subscribed_params.end())
{
  g_params[clean_key] = v;
}
```

should be enough to fix that.

By the way, any chance to get this backported to Hydro?

**Comment by dirk-thomas** [comment]
Thanks for catching this. I have added https://github.com/ros/ros_comm/commit/dcf7ad4023c488c43d74b224c2ba36bde3173397 to the PR.

Since the PR actually changes the behavior I usually prefer to not backport it (especially since Hydro has already being released 8 month ago). But I see the need to have the fix for Hydro as well. Lets merge it for Indigo first, get the change released and used for a few weeks and then consider backporting it?

**Comment by RainCT** [comment]
Sure, that sounds like a plan.

**Comment by efernandez** [comment]
@dirk-thomas Any update on backporting this to hydro?

**Comment by v-lopez** [comment]
@dirk-thomas +1 to @efernandez and @RainCT  request, can we backport this to hydro?

**Comment by dirk-thomas** [comment]
Cherry picked 67a07f10f3c8e9f17441e42e9a2c8741d1a6c3cd 7016e0edaf22ad284eff636376abc57b0c4c9a09 433eab61f64c2a40a191eef66fae081be48d8378. Will be available in 1.10.11 soonish.
